### ISSUE TITLE
Decoding ogg/opus in parallel gets significantly slower with more workers

### ISSUE THREAD

**Issue by sonovice** [eshaz/wasm-audio-decoders#87]
I am trying to decode multiple ogg/opus files in parallel. If using 3 or less workers, this works in reasonable time for my file (~2.5 seconds). When I go for 4, 5 or even more workers, processing times are jumping to 15 seconds and more. Does anyone have a clue what could cause this bottleneck?

This is the start of my custom audio player code:
```typescript
import {OggOpusDecoderWebWorker, OpusDecodedAudio} from 'ogg-opus-decoder';

export class AudioPlayer {
    private audioContext: AudioContext;
    private trackBuffers?: AudioBuffer[];

    constructor() {
        this.audioContext = new AudioContext();
    }

    private loadTrackToAudioBuffer = (trackUrl: string): Promise<AudioBuffer> => {
        const decoder = new OggOpusDecoderWebWorker();
        const decoderPromise = decoder.ready;
        const fetchPromise = fetch(trackUrl)
            .then(response => response.arrayBuffer())
            .then(buffer => new Uint8Array(buffer));

        return Promise.all([decoderPromise, fetchPromise])
            .then(([, array]) => {
                console.log('loaded');
                return decoder.decode(array);
            })
            .then((audio: OpusDecodedAudio) => {
                console.log(`decoded`);
                const buffer = this.audioContext.createBuffer(
                    audio.channelData.length,
                    audio.samplesDecoded,
                    audio.sampleRate
                );
                audio.channelData.forEach((channel, idx) =>
                    buffer.getChannelData(idx).set(channel)
                );
                return buffer;
            })
            .catch(error => {
                throw new Error(`Failed to load track from ${trackUrl}: ${error.message}`);
            })
            .finally(() => {
                decoder.free();
            });
    }

    async loadPlaylist(trackUrls: string[]): Promise<number[]> {
        let trackPromises: Promise<AudioBuffer>[] = trackUrls.map(this.loadTrackToAudioBuffer);
        return Promise.all(trackPromises)
            .then((buffers: AudioBuffer[]) => {
                this.trackBuffers = buffers;
                return buffers.map((buffer: AudioBuffer) => buffer.duration);
            })
            .catch(error => {
                console.error('Failed to load playlist:', error);
                return [];
            });
    }
}
```

It can be utilized using
```typescript
import {AudioPlayer} from "../audio";

const player = new AudioPlayer();
player.loadPlaylist([
    'https://pb.cantico-app.de/api/files/n81nlo40s2sffjc/hugj3a95s34nqda/sample2_ypgz61Tdaw.opus',
    'https://pb.cantico-app.de/api/files/n81nlo40s2sffjc/hugj3a95s34nqda/sample2_ypgz61Tdaw.opus',
    'https://pb.cantico-app.de/api/files/n81nlo40s2sffjc/hugj3a95s34nqda/sample2_ypgz61Tdaw.opus',
    'https://pb.cantico-app.de/api/files/n81nlo40s2sffjc/hugj3a95s34nqda/sample2_ypgz61Tdaw.opus',
]
```

**Comment by eshaz** [comment]
This might happen depending on how many CPU cores you have on your test system. The concurrency benefits stop when you spawn more workers than there are CPU cores. Each additional worker over the number of CPU cores cannot be done in parallel, and has to share CPU time with another task. Also, each additional worker adds a bit of overhead caused by task switching which would slow down the overall decoding.

I would recommend spawning `cpu_cores - 1` workers for the best performance. You can get the number of cores using [`navigator.hardwareConcurrency`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/hardwareConcurrency) on the browser or [`os.cpus().length`](https://stackoverflow.com/a/37871843) in NodeJS.

Your code looks pretty good to me, but you can experiment with parallel decoding on the [`demo page`](https://eshaz.github.io/wasm-audio-decoders/), if you want to rule out your code as being the cause of the slowdown.

**Comment by eshaz** [comment]
Closing due to inactivity. Let me know if you think there is something else happening outside of the explanation I noted above.
